// svelte-language-server/dist/src/lib/documents/TextDocument.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../utils");
var api_1 = require("../../api");
var TextDocument = /** @class */ (function (_super) {
    __extends(TextDocument, _super);
    function TextDocument(url, content) {
        var _this = _super.call(this) || this;
        _this.url = url;
        _this.content = content;
        return _this;
    }
    TextDocument.prototype.getText = function () {
        return this.content;
    };
    TextDocument.prototype.getURL = function () {
        return this.url;
    };
    TextDocument.prototype.setText = function (text) {
        this.content = text;
    };
    TextDocument.prototype.getFilePath = function () {
        return utils_1.urlToPath(this.url);
    };
    TextDocument.prototype.getAttributes = function () {
        return {};
    };
    return TextDocument;
}(api_1.Document));
exports.TextDocument = TextDocument;
