// apollo-language-server/lib/graphqlTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ChangeSeverity;
(function (ChangeSeverity) {
    ChangeSeverity["FAILURE"] = "FAILURE";
    ChangeSeverity["NOTICE"] = "NOTICE";
    ChangeSeverity["WARNING"] = "WARNING";
})(ChangeSeverity = exports.ChangeSeverity || (exports.ChangeSeverity = {}));
var IntrospectionDirectiveLocation;
(function (IntrospectionDirectiveLocation) {
    IntrospectionDirectiveLocation["ARGUMENT_DEFINITION"] = "ARGUMENT_DEFINITION";
    IntrospectionDirectiveLocation["ENUM"] = "ENUM";
    IntrospectionDirectiveLocation["ENUM_VALUE"] = "ENUM_VALUE";
    IntrospectionDirectiveLocation["FIELD"] = "FIELD";
    IntrospectionDirectiveLocation["FIELD_DEFINITION"] = "FIELD_DEFINITION";
    IntrospectionDirectiveLocation["FRAGMENT_DEFINITION"] = "FRAGMENT_DEFINITION";
    IntrospectionDirectiveLocation["FRAGMENT_SPREAD"] = "FRAGMENT_SPREAD";
    IntrospectionDirectiveLocation["INLINE_FRAGMENT"] = "INLINE_FRAGMENT";
    IntrospectionDirectiveLocation["INPUT_FIELD_DEFINITION"] = "INPUT_FIELD_DEFINITION";
    IntrospectionDirectiveLocation["INPUT_OBJECT"] = "INPUT_OBJECT";
    IntrospectionDirectiveLocation["INTERFACE"] = "INTERFACE";
    IntrospectionDirectiveLocation["MUTATION"] = "MUTATION";
    IntrospectionDirectiveLocation["OBJECT"] = "OBJECT";
    IntrospectionDirectiveLocation["QUERY"] = "QUERY";
    IntrospectionDirectiveLocation["SCALAR"] = "SCALAR";
    IntrospectionDirectiveLocation["SCHEMA"] = "SCHEMA";
    IntrospectionDirectiveLocation["SUBSCRIPTION"] = "SUBSCRIPTION";
    IntrospectionDirectiveLocation["UNION"] = "UNION";
})(IntrospectionDirectiveLocation = exports.IntrospectionDirectiveLocation || (exports.IntrospectionDirectiveLocation = {}));
var IntrospectionTypeKind;
(function (IntrospectionTypeKind) {
    IntrospectionTypeKind["ENUM"] = "ENUM";
    IntrospectionTypeKind["INPUT_OBJECT"] = "INPUT_OBJECT";
    IntrospectionTypeKind["INTERFACE"] = "INTERFACE";
    IntrospectionTypeKind["LIST"] = "LIST";
    IntrospectionTypeKind["NON_NULL"] = "NON_NULL";
    IntrospectionTypeKind["OBJECT"] = "OBJECT";
    IntrospectionTypeKind["SCALAR"] = "SCALAR";
    IntrospectionTypeKind["UNION"] = "UNION";
})(IntrospectionTypeKind = exports.IntrospectionTypeKind || (exports.IntrospectionTypeKind = {}));
var ValidationErrorCode;
(function (ValidationErrorCode) {
    ValidationErrorCode["DEPRECATED_FIELD"] = "DEPRECATED_FIELD";
    ValidationErrorCode["INVALID_OPERATION"] = "INVALID_OPERATION";
    ValidationErrorCode["NON_PARSEABLE_DOCUMENT"] = "NON_PARSEABLE_DOCUMENT";
})(ValidationErrorCode = exports.ValidationErrorCode || (exports.ValidationErrorCode = {}));
var ValidationErrorType;
(function (ValidationErrorType) {
    ValidationErrorType["FAILURE"] = "FAILURE";
    ValidationErrorType["INVALID"] = "INVALID";
    ValidationErrorType["WARNING"] = "WARNING";
})(ValidationErrorType = exports.ValidationErrorType || (exports.ValidationErrorType = {}));
