// apollo-language-server/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var validation_1 = require("./errors/validation");
exports.getValidationErrors = validation_1.getValidationErrors;
var logger_1 = require("./errors/logger");
exports.ToolError = logger_1.ToolError;
var base_1 = require("./project/base");
exports.GraphQLProject = base_1.GraphQLProject;
var client_1 = require("./project/client");
exports.isClientProject = client_1.isClientProject;
exports.GraphQLClientProject = client_1.GraphQLClientProject;
var service_1 = require("./project/service");
exports.isServiceProject = service_1.isServiceProject;
exports.GraphQLServiceProject = service_1.GraphQLServiceProject;
var schema_1 = require("./providers/schema");
exports.schemaProviderFromConfig = schema_1.schemaProviderFromConfig;
__export(require("./engine"));
__export(require("./config"));
const graphqlTypes = __importStar(require("./graphqlTypes"));
exports.graphqlTypes = graphqlTypes;
