// ts-lit-plugin/lib/ts-lit-plugin/translate/translate-definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ts_module_1 = require("../../ts-module");
var translate_range_1 = require("./translate-range");
function translateDefinition(definition) {
    var targetNode = definition.target.node;
    var targetStart = targetNode.getStart();
    var targetEnd = targetNode.getEnd();
    var targetFileName = targetNode.getSourceFile().fileName;
    var target = definition.target;
    return {
        definitions: [
            {
                name: ("name" in target && target.name) || ("propName" in target && target.propName) || ("attrName" in target && target.attrName) || "",
                textSpan: {
                    start: targetStart,
                    length: targetEnd - targetStart
                },
                fileName: targetFileName,
                containerName: targetFileName,
                kind: ts_module_1.tsModule.ts.ScriptElementKind.memberVariableElement,
                containerKind: ts_module_1.tsModule.ts.ScriptElementKind.functionElement
            }
        ],
        textSpan: translate_range_1.translateRange(definition.fromRange)
    };
}
exports.translateDefinition = translateDefinition;
