// ts-lit-plugin/lib/ts-lit-plugin/translate/translate-completions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var translate_target_kind_1 = require("./translate-target-kind");
var translate_range_1 = require("./translate-range");
function translateCompletions(completions) {
    var entries = completions.map(function (completion) { return translateCompletion(completion); });
    if (entries != null && entries.length > 0) {
        return {
            isGlobalCompletion: false,
            isMemberCompletion: false,
            isNewIdentifierLocation: false,
            entries: entries
        };
    }
    return undefined;
}
exports.translateCompletions = translateCompletions;
function translateCompletion(completion) {
    var importance = completion.importance, kind = completion.kind, insert = completion.insert, name = completion.name, range = completion.range;
    return __assign({ name: name, kind: translate_target_kind_1.translateTargetKind(kind), kindModifiers: completion.kindModifiers, sortText: importance === "high" ? "0" : importance === "medium" ? "1" : "2", insertText: insert }, (range != null ? { replacementSpan: translate_range_1.translateRange(range) } : {}));
}
