// apollo-language-server/src/providers/federation-info/endpoint.ts
import { execute as linkExecute, toPromise, from } from "apollo-link";
import { createHttpLink, HttpLink } from "apollo-link-http";
import { ExecutionResult, parse } from "graphql";
import { Agent } from "http";
import { fetch } from "apollo-env";
import { RemoteServiceConfig } from "../../config";
import { ApolloFederationInfoProvider, FederationInfo } from "./base";

export class EndpointFederationInfoProvider
  implements ApolloFederationInfoProvider {
  private info?: FederationInfo;

  constructor(private config: Exclude<RemoteServiceConfig, "name">) {}

  private async getFederationInfo() {
    const { skipSSLValidation, url, headers } = this.config;
    const options: HttpLink.Options = {
      uri: url,
      fetch,
      ...(skipSSLValidation && { fetchOptions: { agent: new Agent() } })
    };

    const getFederationInfoQuery = `
      query getFederationInfo {
        _service {
          sdl
        }
      }
    `;

    const { data, errors } = (await toPromise(
      linkExecute(createHttpLink(options), {
        query: parse(getFederationInfoQuery),
        context: { headers }
      })
    )) as ExecutionResult<{ _service: FederationInfo }>;

    if (errors && errors.length) {
      // XXX better error handling of GraphQL errors
      throw new Error(errors.map(({ message }: Error) => message).join("\n"));
    }

    if (!data || !data._service) {
      throw new Error(
        "No data received from server when querying for _service."
      );
    }

    return data._service;
  }

  async resolveFederationInfo() {
    if (!this.info) {
      this.info = await this.getFederationInfo();
    }

    if (!this.info) throw new Error("No service info returned by service");

    return this.info;
  }
}
