// svelte/motion.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var store = require('./store');
var internal = require('./internal');
var easing = require('./easing');

function is_date(obj) {
	return Object.prototype.toString.call(obj) === '[object Date]';
}

function get_initial_velocity(value) {
	if (typeof value === 'number' || is_date(value)) return 0;

	if (Array.isArray(value)) return value.map(get_initial_velocity);

	if (value && typeof value === 'object') {
		const velocities = {};
		for (const k in value) velocities[k] = get_initial_velocity(value[k]);
		return velocities;
	}

	throw new Error(`Cannot spring ${typeof value} values`);
}

function get_threshold(value, target_value, precision) {
	if (typeof value === 'number' || is_date(value)) return precision * Math.abs((target_value - value));

	if (Array.isArray(value)) return value.map((v, i) => get_threshold(v, target_value[i], precision));

	if (value && typeof value === 'object') {
		const threshold = {};
		for (const k in value) threshold[k] = get_threshold(value[k], target_value[k], precision);
		return threshold;
	}

	throw new Error(`Cannot spring ${typeof value} values`);
}

function tick_spring(velocity, current_value, target_value, stiffness, damping, multiplier, threshold) {
	let settled = true;
	let value;

	if (typeof current_value === 'number' || is_date(current_value)) {
		const delta = target_value - current_value;
		const spring = stiffness * delta;
		const damper = damping * velocity;

		const acceleration = spring - damper;

		velocity += acceleration;
		const d = velocity * multiplier;

		if (is_date(current_value)) {
			value = new Date(current_value.getTime() + d);
		} else {
			value = current_value + d;
		}

		if (Math.abs(d) > threshold || Math.abs(delta) > threshold) settled = false;
	}

	else if (Array.isArray(current_value)) {
		value = current_value.map((v, i) => {
			const result = tick_spring(
				velocity[i],
				v,
				target_value[i],
				stiffness,
				damping,
				multiplier,
				threshold[i]
			);

			velocity[i] = result.velocity;
			if (!result.settled) settled = false;
			return result.value;
		});
	}

	else if (typeof current_value === 'object') {
		value = {};
		for (const k in current_value) {
			const result = tick_spring(
				velocity[k],
				current_value[k],
				target_value[k],
				stiffness,
				damping,
				multiplier,
				threshold[k]
			);

			velocity[k] = result.velocity;
			if (!result.settled) settled = false;
			value[k] = result.value;
		}
	}

	else {
		throw new Error(`Cannot spring ${typeof value} values`);
	}

	return { velocity, value, settled };
}

function spring(value, opts = {}) {
	const store$$1 = store.writable(value);

	const { stiffness = 0.15, damping = 0.8, precision = 0.001 } = opts;
	const velocity = get_initial_velocity(value);

	let task;
	let target_value = value;
	let last_time;
	let settled;
	let threshold;
	let current_token;

	function set(new_value) {
		target_value = new_value;
		threshold = get_threshold(value, target_value, spring.precision);

		const token = current_token = {};

		if (!task) {
			last_time = window.performance.now();
			settled = false;

			task = internal.loop(now => {
				({ value, settled } = tick_spring(
					velocity,
					value,
					target_value,
					spring.stiffness,
					spring.damping,
					(now - last_time) * 60 / 1000,
					threshold
				));

				last_time = now;

				if (settled) {
					value = target_value;
					task = null;
				}

				store$$1.set(value);
				return !settled;
			});
		}

		return new Promise(fulfil => {
			task.promise.then(() => {
				if (token === current_token) fulfil();
			});
		});
	}

	const spring = {
		set,
		update: fn => set(fn(target_value, value)),
		subscribe: store$$1.subscribe,
		stiffness,
		damping,
		precision
	};

	return spring;
}

function get_interpolator(a, b) {
	if (a === b || a !== a) return () => a;

	const type = typeof a;

	if (type !== typeof b || Array.isArray(a) !== Array.isArray(b)) {
		throw new Error('Cannot interpolate values of different type');
	}

	if (Array.isArray(a)) {
		const arr = b.map((bi, i) => {
			return get_interpolator(a[i], bi);
		});

		return t => arr.map(fn => fn(t));
	}

	if (type === 'object') {
		if (!a || !b) throw new Error('Object cannot be null');

		if (is_date(a) && is_date(b)) {
			a = a.getTime();
			b = b.getTime();
			const delta = b - a;
			return t => new Date(a + t * delta);
		}

		const keys = Object.keys(b);
		const interpolators = {};

		keys.forEach(key => {
			interpolators[key] = get_interpolator(a[key], b[key]);
		});

		return t => {
			const result = {};
			keys.forEach(key => {
				result[key] = interpolators[key](t);
			});
			return result;
		};
	}

	if (type === 'number') {
		const delta = b - a;
		return t => a + t * delta;
	}

	throw new Error(`Cannot interpolate ${type} values`);
}

function tweened(value, defaults = {}) {
	const store$$1 = store.writable(value);

	let task;
	let target_value = value;

	function set(new_value, opts) {
		target_value = new_value;

		let previous_task = task;
		let started = false;

		let {
			delay = 0,
			duration = 400,
			easing: easing$$1 = easing.linear,
			interpolate = get_interpolator
		} = internal.assign(internal.assign({}, defaults), opts);

		const start = window.performance.now() + delay;
		let fn;

		task = internal.loop(now => {
			if (now < start) return true;

			if (!started) {
				fn = interpolate(value, new_value);
				if (typeof duration === 'function') duration = duration(value, new_value);
				started = true;
			}

			if (previous_task) {
				previous_task.abort();
				previous_task = null;
			}

			const elapsed = now - start;

			if (elapsed > duration) {
				store$$1.set(value = new_value);
				return false;
			}

			store$$1.set(value = fn(easing$$1(elapsed / duration)));
			return true;
		});

		return task.promise;
	}

	return {
		set,
		update: (fn, opts) => set(fn(target_value, value), opts),
		subscribe: store$$1.subscribe
	};
}

exports.spring = spring;
exports.tweened = tweened;
