// vscode-languageserver-protocol/lib/main.js
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
'use strict';
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const vscode_jsonrpc_1 = require("vscode-jsonrpc");
exports.ErrorCodes = vscode_jsonrpc_1.ErrorCodes;
exports.ResponseError = vscode_jsonrpc_1.ResponseError;
exports.CancellationToken = vscode_jsonrpc_1.CancellationToken;
exports.CancellationTokenSource = vscode_jsonrpc_1.CancellationTokenSource;
exports.Disposable = vscode_jsonrpc_1.Disposable;
exports.Event = vscode_jsonrpc_1.Event;
exports.Emitter = vscode_jsonrpc_1.Emitter;
exports.Trace = vscode_jsonrpc_1.Trace;
exports.SetTraceNotification = vscode_jsonrpc_1.SetTraceNotification;
exports.LogTraceNotification = vscode_jsonrpc_1.LogTraceNotification;
exports.RequestType = vscode_jsonrpc_1.RequestType;
exports.RequestType0 = vscode_jsonrpc_1.RequestType0;
exports.NotificationType = vscode_jsonrpc_1.NotificationType;
exports.NotificationType0 = vscode_jsonrpc_1.NotificationType0;
exports.MessageReader = vscode_jsonrpc_1.MessageReader;
exports.MessageWriter = vscode_jsonrpc_1.MessageWriter;
exports.ConnectionStrategy = vscode_jsonrpc_1.ConnectionStrategy;
exports.StreamMessageReader = vscode_jsonrpc_1.StreamMessageReader;
exports.StreamMessageWriter = vscode_jsonrpc_1.StreamMessageWriter;
exports.IPCMessageReader = vscode_jsonrpc_1.IPCMessageReader;
exports.IPCMessageWriter = vscode_jsonrpc_1.IPCMessageWriter;
exports.createClientPipeTransport = vscode_jsonrpc_1.createClientPipeTransport;
exports.createServerPipeTransport = vscode_jsonrpc_1.createServerPipeTransport;
exports.generateRandomPipeName = vscode_jsonrpc_1.generateRandomPipeName;
exports.createClientSocketTransport = vscode_jsonrpc_1.createClientSocketTransport;
exports.createServerSocketTransport = vscode_jsonrpc_1.createServerSocketTransport;
__export(require("vscode-languageserver-types"));
__export(require("./protocol"));
const config = require("./protocol.configuration.proposed");
const folders = require("./protocol.workspaceFolders.proposed");
const color = require("./protocol.colorProvider.proposed");
var Proposed;
(function (Proposed) {
    let ConfigurationRequest;
    (function (ConfigurationRequest) {
        ConfigurationRequest.type = config.ConfigurationRequest.type;
    })(ConfigurationRequest = Proposed.ConfigurationRequest || (Proposed.ConfigurationRequest = {}));
    ;
    let WorkspaceFoldersRequest;
    (function (WorkspaceFoldersRequest) {
        WorkspaceFoldersRequest.type = folders.WorkspaceFoldersRequest.type;
    })(WorkspaceFoldersRequest = Proposed.WorkspaceFoldersRequest || (Proposed.WorkspaceFoldersRequest = {}));
    let DidChangeWorkspaceFoldersNotification;
    (function (DidChangeWorkspaceFoldersNotification) {
        DidChangeWorkspaceFoldersNotification.type = folders.DidChangeWorkspaceFoldersNotification.type;
    })(DidChangeWorkspaceFoldersNotification = Proposed.DidChangeWorkspaceFoldersNotification || (Proposed.DidChangeWorkspaceFoldersNotification = {}));
    Proposed.DocumentColorRequest = color.DocumentColorRequest;
    Proposed.ColorPresentationRequest = color.ColorPresentationRequest;
})(Proposed = exports.Proposed || (exports.Proposed = {}));
function createProtocolConnection(reader, writer, logger, strategy) {
    return vscode_jsonrpc_1.createMessageConnection(reader, writer, logger, strategy);
}
exports.createProtocolConnection = createProtocolConnection;
