// typescript-template-language-service-decorator/lib/standard-script-source-helper.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
var nodes_1 = require("./nodes");
var StandardScriptSourceHelper = /** @class */ (function () {
    function StandardScriptSourceHelper(typescript, project) {
        this.typescript = typescript;
        this.project = project;
    }
    StandardScriptSourceHelper.prototype.getNode = function (fileName, position) {
        var s = this.getProgram().getSourceFile(fileName);
        return s && nodes_1.findNode(this.typescript, s, position);
    };
    StandardScriptSourceHelper.prototype.getAllNodes = function (fileName, cond) {
        var s = this.getProgram().getSourceFile(fileName);
        return s ? nodes_1.findAllNodes(this.typescript, s, cond) : [];
    };
    StandardScriptSourceHelper.prototype.getLineAndChar = function (fileName, position) {
        var s = this.project.getScriptInfo(fileName);
        var location = s.positionToLineOffset(position);
        return { line: location.line - 1, character: location.offset - 1 };
    };
    StandardScriptSourceHelper.prototype.getOffset = function (fileName, line, character) {
        var s = this.project.getScriptInfo(fileName);
        return s.lineOffsetToPosition(line + 1, character + 1);
    };
    StandardScriptSourceHelper.prototype.getProgram = function () {
        return this.project.getLanguageService().getProgram();
    };
    return StandardScriptSourceHelper;
}());
exports.default = StandardScriptSourceHelper;
