// @prettier/plugin-pug/dist/options/attribute-separator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveAttributeSeparatorOption = exports.PUG_ATTRIBUTE_SEPARATOR_OPTION = exports.ATTRIBUTE_SEPARATOR_OPTION = void 0;
const _1 = require(".");
exports.ATTRIBUTE_SEPARATOR_OPTION = {
    since: '1.0.0',
    category: _1.CATEGORY_PUG,
    type: 'choice',
    default: 'always',
    description: 'Change when attributes are separated by commas in tags.',
    choices: [
        {
            value: 'always',
            description: 'Always separate attributes with commas. Example: `button(type="submit", (click)="play()", disabled)`'
        },
        {
            value: 'as-needed',
            description: 'Only add commas between attributes where required. Example: `button(type="submit", (click)="play()" disabled)`'
        }
    ]
};
exports.PUG_ATTRIBUTE_SEPARATOR_OPTION = {
    ...exports.ATTRIBUTE_SEPARATOR_OPTION,
    since: '1.6.0',
    default: null,
    choices: [
        {
            value: null,
            description: 'Use `attributeSeparator` value.'
        },
        {
            value: 'always',
            description: 'Always separate attributes with commas. Example: `button(type="submit", (click)="play()", disabled)`'
        },
        {
            value: 'as-needed',
            description: 'Only add commas between attributes where required. Example: `button(type="submit", (click)="play()" disabled)`'
        }
    ]
};
function resolveAttributeSeparatorOption(attributeSeparator) {
    switch (attributeSeparator) {
        case 'always':
            return true;
        case 'as-needed':
            return false;
    }
    throw new Error(`Invalid option for pug attributeSeparator. Found '${attributeSeparator}'. Possible options: 'always' or 'as-needed'`);
}
exports.resolveAttributeSeparatorOption = resolveAttributeSeparatorOption;
