// didyoumean2/dist/lib/fillDefaultOptions.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var R = __importStar(require("ramda"));
var ReturnTypeEnums_1 = require("../enums/ReturnTypeEnums");
var ThresholdTypeEnums_1 = require("../enums/ThresholdTypeEnums");
var errors_1 = require("../errors");
var fillDefaultOptions = R.compose(R.cond([
    [
        R.propEq('thresholdType', ThresholdTypeEnums_1.ThresholdTypeEnums.EDIT_DISTANCE),
        R.merge({
            threshold: 20
        })
    ],
    [
        R.propEq('thresholdType', ThresholdTypeEnums_1.ThresholdTypeEnums.SIMILARITY),
        R.merge({
            threshold: 0.4
        })
    ],
    [
        R.T,
        /* istanbul ignore next */
        function () {
            throw errors_1.unknownThresholdTypeError;
        }
    ]
]), R.merge({
    caseSensitive: false,
    deburr: false,
    matchPath: [],
    returnType: ReturnTypeEnums_1.ReturnTypeEnums.FIRST_CLOSEST_MATCH,
    thresholdType: ThresholdTypeEnums_1.ThresholdTypeEnums.SIMILARITY,
    trimSpaces: true
}), R.defaultTo({}));
exports.default = fillDefaultOptions;
