// ts-doc-gen/dist/messages.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var chalk_1 = __importDefault(require("chalk"));
var log_1 = require("@sorg/log");
exports.successFile = function (filePath) {
    console.log(chalk_1.default.green.bold("Done! File created at " + filePath));
};
exports.success = function (message) {
    console.log(chalk_1.default.green.bold(message));
};
exports.failure = function (message) {
    console.log(chalk_1.default.red.bold(message));
};
exports.info = function (message) {
    console.log(chalk_1.default.yellow.bold(message));
};
exports.logger = new log_1.Logger({
    help: {
        styles: [
            { color: '#72a', background: '#111' },
            { color: '#f23', background: '#222' },
            { color: '#2af', background: '#222' }
        ],
        preset: new log_1.PresetNodeHelp("--out: OUTPUT_FILE:Relative path to the output file with extension.\n--dir: DIR:        Relative path to the input dir.\n--name: NAME:       Header name.\n--exclude: FILES:      Comma separated list of files to exclude, cannot be used with --include\n--include: FILES:      Comma separated list of files to include, cannot be used with --exclude\n-h: :           Displays this Message.", ':', 11, 90)
    }
});
