// typescript-template-language-service-decorator/lib/standard-script-source-helper.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
var nodes_1 = require("./nodes");
var StandardScriptSourceHelper = /** @class */ (function () {
    function StandardScriptSourceHelper(typescript, languageService) {
        this.typescript = typescript;
        this.languageService = languageService;
    }
    StandardScriptSourceHelper.prototype.getNode = function (fileName, position) {
        var s = this.languageService.getProgram().getSourceFile(fileName);
        return s && nodes_1.findNode(this.typescript, s, position);
    };
    StandardScriptSourceHelper.prototype.getAllNodes = function (fileName, cond) {
        var s = this.languageService.getProgram().getSourceFile(fileName);
        return s ? nodes_1.findAllNodes(this.typescript, s, cond) : [];
    };
    StandardScriptSourceHelper.prototype.getLineAndChar = function (fileName, position) {
        var s = this.languageService.getProgram().getSourceFile(fileName);
        return this.typescript.getLineAndCharacterOfPosition(s, position);
    };
    StandardScriptSourceHelper.prototype.getOffset = function (fileName, line, character) {
        var s = this.languageService.getProgram().getSourceFile(fileName);
        return this.typescript.getPositionOfLineAndCharacter(s, line, character);
    };
    return StandardScriptSourceHelper;
}());
exports.default = StandardScriptSourceHelper;
