// vscode-jsonrpc/lib/common/api.js
"use strict";
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
/// <reference path="../../typings/thenable.d.ts" />
Object.defineProperty(exports, "__esModule", { value: true });
const messages_1 = require("../common/messages");
exports.RequestType = messages_1.RequestType;
exports.RequestType0 = messages_1.RequestType0;
exports.RequestType1 = messages_1.RequestType1;
exports.RequestType2 = messages_1.RequestType2;
exports.RequestType3 = messages_1.RequestType3;
exports.RequestType4 = messages_1.RequestType4;
exports.RequestType5 = messages_1.RequestType5;
exports.RequestType6 = messages_1.RequestType6;
exports.RequestType7 = messages_1.RequestType7;
exports.RequestType8 = messages_1.RequestType8;
exports.RequestType9 = messages_1.RequestType9;
exports.ResponseError = messages_1.ResponseError;
exports.ErrorCodes = messages_1.ErrorCodes;
exports.NotificationType = messages_1.NotificationType;
exports.NotificationType0 = messages_1.NotificationType0;
exports.NotificationType1 = messages_1.NotificationType1;
exports.NotificationType2 = messages_1.NotificationType2;
exports.NotificationType3 = messages_1.NotificationType3;
exports.NotificationType4 = messages_1.NotificationType4;
exports.NotificationType5 = messages_1.NotificationType5;
exports.NotificationType6 = messages_1.NotificationType6;
exports.NotificationType7 = messages_1.NotificationType7;
exports.NotificationType8 = messages_1.NotificationType8;
exports.NotificationType9 = messages_1.NotificationType9;
exports.ParameterStructures = messages_1.ParameterStructures;
const disposable_1 = require("../common/disposable");
exports.Disposable = disposable_1.Disposable;
const events_1 = require("../common/events");
exports.Event = events_1.Event;
exports.Emitter = events_1.Emitter;
const cancellation_1 = require("../common/cancellation");
exports.CancellationTokenSource = cancellation_1.CancellationTokenSource;
exports.CancellationToken = cancellation_1.CancellationToken;
const messageReader_1 = require("../common/messageReader");
exports.MessageReader = messageReader_1.MessageReader;
exports.AbstractMessageReader = messageReader_1.AbstractMessageReader;
exports.ReadableStreamMessageReader = messageReader_1.ReadableStreamMessageReader;
const messageWriter_1 = require("../common/messageWriter");
exports.MessageWriter = messageWriter_1.MessageWriter;
exports.AbstractMessageWriter = messageWriter_1.AbstractMessageWriter;
exports.WriteableStreamMessageWriter = messageWriter_1.WriteableStreamMessageWriter;
const connection_1 = require("../common/connection");
exports.ConnectionStrategy = connection_1.ConnectionStrategy;
exports.ConnectionOptions = connection_1.ConnectionOptions;
exports.NullLogger = connection_1.NullLogger;
exports.createMessageConnection = connection_1.createMessageConnection;
exports.ProgressType = connection_1.ProgressType;
exports.Trace = connection_1.Trace;
exports.TraceFormat = connection_1.TraceFormat;
exports.SetTraceNotification = connection_1.SetTraceNotification;
exports.LogTraceNotification = connection_1.LogTraceNotification;
exports.ConnectionErrors = connection_1.ConnectionErrors;
exports.ConnectionError = connection_1.ConnectionError;
exports.CancellationReceiverStrategy = connection_1.CancellationReceiverStrategy;
exports.CancellationSenderStrategy = connection_1.CancellationSenderStrategy;
exports.CancellationStrategy = connection_1.CancellationStrategy;
const ral_1 = require("./ral");
exports.RAL = ral_1.default;
