// sass-formatter/dist/state.js
"use strict";
exports.__esModule = true;
exports.FormattingState = void 0;
var FormattingState = /** @class */ (function () {
    function FormattingState() {
        this.lines = [];
        /** Current line index. */
        this.currentLine = 0;
        /** Formatting Result */
        this.RESULT = '';
        /** Context For Each Line. */
        this.LOCAL_CONTEXT = {
            isAdjacentSelector: false,
            isHtmlTag: false,
            ResetTabs: false,
            indentation: {
                distance: 0,
                offset: 0
            },
            isAtExtend: false,
            isAnd_: false,
            isClassOrIdSelector: false,
            isIfOrElseAProp: false,
            isAtKeyframes: false,
            isAtKeyframesPoint: false,
            isProp: false,
            isInterpolatedProp: false,
            isInclude: false
        };
        this.CONTEXT = {
            blockCommentDistance: 0,
            isFirstLine: true,
            isLastLine: false,
            allowSpace: false,
            isInBlockComment: false,
            ignoreLine: false,
            lastSelectorTabs: 0,
            wasLastLineSelector: false,
            convert: {
                lastSelector: '',
                wasLastLineCss: false
            },
            keyframes: {
                is: false,
                tabs: 0
            },
            tabs: 0,
            currentTabs: 0,
            firstCommaHeader: { exists: false, distance: 0 }
        };
        this.CONFIG = {
            insertSpaces: true,
            tabSize: 2,
            convert: true,
            debug: false,
            deleteEmptyRows: true,
            deleteWhitespace: true,
            setPropertySpace: true
        };
    }
    FormattingState.prototype.setLocalContext = function (context) {
        this.LOCAL_CONTEXT = context;
    };
    return FormattingState;
}());
exports.FormattingState = FormattingState;
