// svelte-language-server/dist/src/api/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vscode_languageserver_types_1 = require("vscode-languageserver-types");
exports.Diagnostic = vscode_languageserver_types_1.Diagnostic;
exports.DiagnosticSeverity = vscode_languageserver_types_1.DiagnosticSeverity;
exports.Position = vscode_languageserver_types_1.Position;
exports.Range = vscode_languageserver_types_1.Range;
exports.Hover = vscode_languageserver_types_1.Hover;
exports.MarkupContent = vscode_languageserver_types_1.MarkupContent;
exports.MarkedString = vscode_languageserver_types_1.MarkedString;
exports.CompletionItem = vscode_languageserver_types_1.CompletionItem;
exports.CompletionItemKind = vscode_languageserver_types_1.CompletionItemKind;
exports.TextEdit = vscode_languageserver_types_1.TextEdit;
exports.InsertTextFormat = vscode_languageserver_types_1.InsertTextFormat;
exports.Command = vscode_languageserver_types_1.Command;
exports.CompletionList = vscode_languageserver_types_1.CompletionList;
exports.TextDocumentItem = vscode_languageserver_types_1.TextDocumentItem;
exports.ColorInformation = vscode_languageserver_types_1.ColorInformation;
exports.ColorPresentation = vscode_languageserver_types_1.ColorPresentation;
exports.Color = vscode_languageserver_types_1.Color;
exports.SymbolInformation = vscode_languageserver_types_1.SymbolInformation;
exports.Location = vscode_languageserver_types_1.Location;
exports.SymbolKind = vscode_languageserver_types_1.SymbolKind;
var DiagnosticsProvider;
(function (DiagnosticsProvider) {
    function is(obj) {
        return typeof obj.getDiagnostics === 'function';
    }
    DiagnosticsProvider.is = is;
})(DiagnosticsProvider = exports.DiagnosticsProvider || (exports.DiagnosticsProvider = {}));
var HoverProvider;
(function (HoverProvider) {
    function is(obj) {
        return typeof obj.doHover === 'function';
    }
    HoverProvider.is = is;
})(HoverProvider = exports.HoverProvider || (exports.HoverProvider = {}));
var CompletionsProvider;
(function (CompletionsProvider) {
    function is(obj) {
        return typeof obj.getCompletions === 'function';
    }
    CompletionsProvider.is = is;
})(CompletionsProvider = exports.CompletionsProvider || (exports.CompletionsProvider = {}));
var FormattingProvider;
(function (FormattingProvider) {
    function is(obj) {
        return typeof obj.formatDocument === 'function';
    }
    FormattingProvider.is = is;
})(FormattingProvider = exports.FormattingProvider || (exports.FormattingProvider = {}));
var TagCompleteProvider;
(function (TagCompleteProvider) {
    function is(obj) {
        return typeof obj.doTagComplete === 'function';
    }
    TagCompleteProvider.is = is;
})(TagCompleteProvider = exports.TagCompleteProvider || (exports.TagCompleteProvider = {}));
var DocumentColorsProvider;
(function (DocumentColorsProvider) {
    function is(obj) {
        return typeof obj.getDocumentColors === 'function';
    }
    DocumentColorsProvider.is = is;
})(DocumentColorsProvider = exports.DocumentColorsProvider || (exports.DocumentColorsProvider = {}));
var ColorPresentationsProvider;
(function (ColorPresentationsProvider) {
    function is(obj) {
        return typeof obj.getColorPresentations === 'function';
    }
    ColorPresentationsProvider.is = is;
})(ColorPresentationsProvider = exports.ColorPresentationsProvider || (exports.ColorPresentationsProvider = {}));
var DocumentSymbolsProvider;
(function (DocumentSymbolsProvider) {
    function is(obj) {
        return typeof obj.getDocumentSymbols === 'function';
    }
    DocumentSymbolsProvider.is = is;
})(DocumentSymbolsProvider = exports.DocumentSymbolsProvider || (exports.DocumentSymbolsProvider = {}));
