// elm-module-parser/dist/src/module_parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
const elmModuleParser = util_1.loadParser('elm_module_parser');
const ElmDefaultImports = elmModuleParser(`
module Default

import Basics exposing (..)
import List exposing (List, (::))
import Maybe exposing (Maybe(..))
import Result exposing (Result(..))
import String exposing (String)
import Char exposing (Char)
import Tuple
import Debug
import Platform exposing ( Program )
import Platform.Cmd as Cmd exposing ( Cmd )
import Platform.Sub as Sub exposing ( Sub )
`).imports.map((a) => ({
    type: 'default-import',
    module: a.module,
    location: null,
    alias: null,
    exposes_all: a.exposes_all,
    exposing: a.exposing.map(e => ({
        type: e.type,
        name: e.name,
        location: null,
    })),
}));
exports.parseElmModule = (input) => {
    const parsed_module = elmModuleParser(`${input}\n`);
    return Object.assign({}, parsed_module, { imports: parsed_module.imports.concat(ElmDefaultImports) });
};
