// typescript-styled-plugin/lib/_virtual-document-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Standard virtual document provider for styled content.
 *
 * Wraps content in a top level `:root { }` rule to make css language service happy
 * since styled allows properties to be top level elements.
 */
class StyledVirtualDocumentFactory {
    createVirtualDocument(context) {
        const contents = `${StyledVirtualDocumentFactory.wrapperPre}${context.text}\n}`;
        return {
            uri: 'untitled://embedded.scss',
            languageId: 'scss',
            version: 1,
            getText: () => contents,
            positionAt: (offset) => {
                const pos = context.toPosition(this.fromVirtualDocOffset(offset));
                return this.toVirtualDocPosition(pos);
            },
            offsetAt: (p) => {
                const offset = context.toOffset(this.fromVirtualDocPosition(p));
                return this.toVirtualDocOffset(offset);
            },
            lineCount: contents.split(/\n/g).length + 1,
        };
    }
    toVirtualDocPosition(position) {
        return {
            line: position.line + 1,
            character: position.character,
        };
    }
    fromVirtualDocPosition(position) {
        return {
            line: position.line - 1,
            character: position.character,
        };
    }
    toVirtualDocOffset(offset) {
        return offset + StyledVirtualDocumentFactory.wrapperPre.length;
    }
    fromVirtualDocOffset(offset) {
        return offset - StyledVirtualDocumentFactory.wrapperPre.length;
    }
}
StyledVirtualDocumentFactory.wrapperPre = ':root{\n';
exports.StyledVirtualDocumentFactory = StyledVirtualDocumentFactory;
